### ISSUE TITLE
Modifing layer data in a callback is not updated in viewer

### ISSUE THREAD

**Issue by Karol-G** [napari/napari#3855]
## 🐛 Bug

If I modify the layer data in a callback than it is not updated in the viewer. I am not sure if this is a bug or if I am missing something.

## To Reproduce

1. Execute code
2. Click on "Activate bug" on right side
3. Move the mouse on the image or perform a click
4. See that the image is not changed
5. Make the image invisible and visible again
6. See that the image is updated

```python
from qtpy import QtWidgets
from napari_plugin_engine import napari_hook_implementation
import napari
import numpy as np

class MyWidget(QtWidgets.QWidget):
    def __init__(self, viewer: 'napari.viewer.Viewer') -> None:
        super().__init__()
        self.viewer = viewer
        self.btn = QtWidgets.QPushButton('Activate bug')
        self.btn.clicked.connect(self._on_click_init)

        self.setLayout(QtWidgets.QVBoxLayout())
        self.layout().addWidget(self.btn)

    def _on_click_init(self):
        my_layer = self.viewer.layers[0]

        @my_layer.mouse_move_callbacks.append
        def do_something1(layer, event):
            layer.data[...] = np.random.randint(low=0, high=255, size=(5, 10, 10))

        @my_layer.mouse_drag_callbacks.append
        def do_something2(layer, event):
            layer.data[...] = np.random.randint(low=0, high=255, size=(5, 10, 10))
            yield
            while event.type == "mouse_move":
                layer.data[...] = np.random.randint(low=0, high=255, size=(5, 10, 10))
                yield


@napari_hook_implementation
def napari_experimental_provide_dock_widget():
    return MyWidget


image = np.random.randint(low=0, high=255, size=(5, 10, 10))
viewer = napari.Viewer()
viewer.window.add_dock_widget(MyWidget(viewer))
viewer.add_image(image)
napari.run()
```

## Expected behavior

Layer data should be updated in the viewer, I guess?

## Environment

napari: 0.4.13.dev134+gfcde7328
Platform: Windows-10-10.0.19041-SP0
Python: 3.8.5 (default, Sep 3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)]
Qt: 5.9.7
PyQt5: 5.9.2
NumPy: 1.21.4
SciPy: 1.7.3
Dask: 2021.12.0
VisPy: 0.9.4

OpenGL:

    GL version: 4.6.0 - Build 27.20.100.8729
    MAX_TEXTURE_SIZE: 16384

Screens:

    screen 1: resolution 1920x1080, scale 1.0

Plugins:

    PartSeg Analysis project reader: 0.13.14
    PartSeg Image reader: 0.13.14
    PartSeg Image tiff writer: 0.13.14
    PartSeg Mask project reader: 0.13.14
    PartSeg Mask project writer: 0.13.14
    PartSeg Masked Tiff reader: 0.13.14
    PartSeg widgets: 0.13.14
    napari-medical-image-formats: 0.3.6
    napari-tools-menu: 0.1.7
    scikit-image: 0.4.13.dev134+gfcde7328
    svg: 0.1.5

Best
Karol

**Comment by jni** [comment]
@Karol-G unfortunately we don't have any way of knowing whether the data has been updated in-place. We've had various [discussions](https://github.com/napari/napari/pull/3155) about wrapping arrays so that `__setitem__` is overridden and we can keep track of things, but unfortunately this doesn't work in various scenarios. For example:

```python
mydata = np.random.random((10, 512, 512))
layer = viewer.add_labels(my_data)  # now setitem on layer.data is tracked
view = layer.data[5]  # this is a view: editing it edits the data in-place
# BUT, the custom setitem is gone!
view[4, 4] = 2  # napari can't know this happened
```

You could say, ok, but what if you override `__getitem__` as well so that views are evented? Unfortunately, that still doesn't work, because the array could be modified totally outside napari:

```python
mydata = np.random.random((10, 512, 512))
layer = viewer.add_labels(my_data)  # now setitem on layer.data is tracked
view = mydata[5]  # this is a view: editing it edits the data in-place
# BUT, napari has no idea about what you do with mydata!
view[4, 4] = 2  # napari can't know this happened
```

In other words, you should use `layer.events.data()` any time you want to *tell* napari that you have changed the data.

I am going to close this as it is more of a fundamental design limitation than a bug, but feel free to keep the discussion going.

**Comment by Karol-G** [comment]
Hmmm, it seems `layer.events.data()` does not really update the viewer either after mouve move or mouse click is called. Only if I zoom in afterwards the image is updated.
